### DISCUSSION TITLE
[RFC] Ability to fully control the name of resources

### DISCUSSION THREAD

**Discussion by mikhailshilkov** [pulumi/pulumi#17592] [Ideas]
Hey folks!

[Ability to fully control the name of resources](https://github.com/pulumi/pulumi/issues/1518) is a highly requested feature where the ask is to be able to customize resource names automatically generated by Pulumi ("autonaming"). We have a design proposal for a lever to configure autonaming, or disable it entirely.

Please have a look at the document below and feel free to share your thoughts with us.

Thanks!

---

## Background

Resources in Pulumi use *autonaming* – a feature whereby Pulumi ensures unique names for “physical” (cloud-side) resources by generating names or name suffixes. For instance, if one uses the AWS provider to write:

```ts
const foo = new aws.rds.Instance(“foo”, { ... });
```

One will end up with an RDS database named (for example) `foo-ab158dc9`. The motivations for this feature are several – being able to handle create-before-replace semantics to aid in various zero-downtime scenarios; being able to spin up multiple instances of the same stack into a single account/cloud-side presence; and so on.

If one wants to specify a resource name exactly, this can nearly always be accomplished by explicitly specifying a specific input (rather than relying on the Pulumi resource name). Often this is called `name`, but this is not always the case; for RDS instances the input is named `identifier`:

```ts
const foo = new aws.rds.Instance(“foo”, { identifier: “foo”, ... });
```

RDS instances also happen to offer an `identifierPrefix` input which allows some degree of control over names whilst still asking Pulumi to generate parts of the name, but this is specific to this resource.

While the current default autonaming approach is to add a dash and hex suffix, a provider can decide to apply other autonaming techniques based on resource type limitations. Here is a list of known deviations:

* Use a different separator that is allowed, e.g. an underscore, or use no separator  
* Convert the name to lowercase  
* Remove all characters other than alphanumeric  
* Generate a suffix of alphabetic character  
* Use the logical name verbatim  
* Generate a UUID  
* Add a fixed prefix (e.g. “alias/” for a KMS alias) or a fixed suffix (e.g. “.fifo” for an SQS queue)  
* Compose a multi-part name combining parent “path” parts (other resource properties) with the name, e.g. projects/{projectName}/resources/{resourceName}

## Problem

However, the default behavior is not always ideal, and many users and customers have wanted broad-spectrum or high-level control over how and when autonaming works for a long time. [https://github.com/pulumi/pulumi/issues/1518](https://github.com/pulumi/pulumi/issues/1518) has been open since June 2018 and seen a lot of discussion and input around different scenarios and problems around moving from other infrastructure-as-code tools.

## Goals

“Control” of naming means many things to many people:

* Disabling autonaming with generated suffixes with “what I write is what I get” semantics, whereby the Pulumi resource name given is that used for the underlying cloud-side resource ([https://github.com/pulumi/pulumi/issues/1518\#issuecomment-683639523](https://github.com/pulumi/pulumi/issues/1518#issuecomment-683639523), [https://github.com/pulumi/pulumi/issues/1518\#issuecomment-748634377](https://github.com/pulumi/pulumi/issues/1518#issuecomment-748634377)).  
* Tweaking autonaming behavior so that it adheres to a set of policies (e.g. snake\_case, camelCase, no longer than X characters, always use UUIDs, etc.) ([https://github.com/pulumi/pulumi/issues/1518\#issuecomment-646594645](https://github.com/pulumi/pulumi/issues/1518#issuecomment-646594645), [https://github.com/pulumi/pulumi/issues/1518\#issuecomment-555017970](https://github.com/pulumi/pulumi/issues/1518#issuecomment-555017970), [https://github.com/pulumi/pulumi/issues/1518\#issuecomment-1146753346](https://github.com/pulumi/pulumi/issues/1518#issuecomment-1146753346), [https://github.com/pulumi/pulumi/issues/1518\#issuecomment-1550849249](https://github.com/pulumi/pulumi/issues/1518#issuecomment-1550849249)).  
* Having different naming strategies for different groups of resource (e.g. by provider, by stack) ([https://github.com/pulumi/pulumi/issues/1518\#issuecomment-1023080341](https://github.com/pulumi/pulumi/issues/1518#issuecomment-1023080341), [https://github.com/pulumi/pulumi/issues/1771](https://github.com/pulumi/pulumi/issues/1771)).

Our understanding is that these categories are listed in the order of relative demand:

1. Most users will keep using our current autonaming approach  
2. A lot of users will disable “name mangling” and want the names to be copied as specified  
3. A smaller subset of users will want autonaming but want to specify their own naming template, potentially customized per provider  
4. A small group of users wants to fine-tune naming of some niche resources with unusual naming requirements

Therefore, (1) should be default, (2) should be very simple, (3) should be straightforward, and (4) should be possible.

## Non-Goals

Autonaming configured in explicit providers, or otherwise in code, is currently out of scope. While we could add extra programmatic APIs, there is no clear case for this yet.

## API Proposal

This section describes the UX of autonaming configuration.

We introduce a new configuration key `pulumi:autonaming`. It can be set on the stack or the project level, or applied across the organization with ESC. It works across all runtimes, including YAML.

```yaml
config:
  pulumi:autonaming: ...
```

Here are some examples of configurations to achieve target scenarios.

### Default autonaming with suffixes

Keep the existing behavior:

```yaml
pulumi:autonaming:
  mode: default
```

It has the same effect as no configuration at all.

### Disable name mangling

Copy the logical name to the physical name as-is:

```yaml
pulumi:autonaming:
  mode: verbatim
```

### Disable autonaming entirely

Never assign resource names at all, require the code to specify them manually as a resource input:

```yaml
pulumi:autonaming:
  mode: disabled
```

### Custom autonaming template

Users can specify a template string to control resource names. The following is similar to our default implementation

```yaml
pulumi:autonaming:
  pattern: ${name}-${hex(8)}
```

Although note that the default currently differs per provider. All the `${exp}` expressions are evaluated at the time of name generation, with the following expressions supported:

| Expression | Description | Example |
| :---- | :---- | :---- |
| name | Logical resource name | ${name} \- same as verbatim |
| hex(n) | Hexadecimal random string of given length | ${name}-${hex(5)} |
| alphanum(n), string(n), num(n) | Alphanumeric string of given length, or just letters or digits | ${name}${num(4)} |
| uuid | UUID | ${uuid} |
| org | Organization name | ${organization}\_${name} |
| project | Project name | ${organization}\_${project}\_${name} |
| stack | Stack name | ${project}\_${stack}\_${name} |
| config.key | Config value of key | ${config.region}\_${name} |

### Configure per provider or per resource type

For advanced scenarios, we support hierarchical config sections

```yaml
pulumi:autonaming:
  mode: default
  providers:
    aws:
      pattern: ${name}_${hex(4)}  # AWS - use underscore and shorter suffix
    azure-native:
      mode: verbatim  # disable name mangling for Azure
      resources:
        "azure-native:storage:Account": ${name}${string(6)}  # storage accounts are globally unique, so add a suffix
```

This configuration changes autonaming approach for AWS and Azure Native providers but keeps the defaults for all other providers.

### Configuration value schema

Here is a JSON schema representation of possible autonaming configurations:

```json
{
  "$schema": "http://json-schema.org/draft-07/schema#",
  "title": "Pulumi Autonaming Configuration",
  "description": "Schema for configuring resource name generation in Pulumi",
  "type": "object",
  "allOf": [
    {
      "$ref": "#/definitions/namingConfig"
    }
  ],
  "properties": {
    "providers": {
      "type": "object",
      "patternProperties": {
        "^[a-zA-Z][a-zA-Z0-9-]*$": {
          "type": "object",
          "allOf": [
            {
              "$ref": "#/definitions/namingConfig"
            }
          ],
          "properties": {
            "resources": {
              "type": "object",
              "patternProperties": {
                "^[a-zA-Z0-9-]+:[a-zA-Z0-9-/]+:[a-zA-Z0-9]+$": {
                  "allOf": [
                    {
                      "$ref": "#/definitions/namingConfig"
                    }
                  ]
                }
              }
            }
          }
        }
      }
    }
  },
  "definitions": {
    "namingConfig": {
      "type": "object",
      "properties": {
        "mode": {
          "type": "string",
          "enum": ["default", "verbatim", "disabled"],
          "description": "Autonaming mode: default (standard Pulumi behavior), verbatim (use logical names), or disabled (require explicit names)"
        },
        "pattern": {
          "type": "string",
          "pattern": "^\\${(?:(?:config\\.[a-zA-Z][a-zA-Z0-9]*)|[a-zA-Z][a-zA-Z0-9]*(?:\\([0-9]+\\))?)}(?:-?\\${(?:(?:config\\.[a-zA-Z][a-zA-Z0-9]*)|[a-zA-Z][a-zA-Z0-9]*(?:\\([0-9]+\\))?)})*$",
          "description": "Template string for custom name generation"
        },
        "enforce": {
          "type": "boolean",
          "description": "When true, prevents providers from modifying the specified naming pattern",
          "default": false
        }
      },
      "oneOf": [
        {
          "required": ["mode"]
        },
        {
          "required": ["pattern"]
        }
      ]
    }
  }
}
```

### User preference vs. provider configuration

A lot of providers specify a custom autonaming configuration for certain resource types (see the **Background** section), for example, use `${name}-${hex(8)}` by default but `lowercase(${name}${hex(8)})` for a resource that only allows lowercase alphanumeric names.

When users choose the default autonaming approach, these limitations still apply. Broadly, we don’t want our users to provide exception lists explicitly when we can adjust the name automatically.

When users provide a custom autonaming template, we want to keep that extra provider knowledge and apply it intelligently. Otherwise, users would have to encode all the same rules in their configurations. To enable this, name generation should consist of two steps:

1. Generate a resource name based on its logical name and user-supplied template.  
2. Transform the result with the provider's rule for the resource type, if any. Transformation rules are described below.

The provider should apply any transformations that barely affect its randomness to user’s template:

* Use a different separator than the one configured, e.g. an underscore, or use no separator  
* Convert the name to lowercase  
* Add a fixed prefix (e.g. “alias/” for a KMS alias) or a fixed suffix (e.g. “.fifo” for an SQS queue)  
* Compose a multi-part name combining parent “path” parts (other resource properties) with the name, e.g. projects/{projectName}/resources/{resourceName}

It can also choose to ignore the user’s template:

* Use the logical name verbatim in places where autonaming is presently off  
* Generate a UUID

The provider should error for cases when a name transformation can dangerously reduce randomness:

* Where the generated name is longer than maximum length allowed (same as today)  
* When illegal characters are used (except dashes, which can be removed)

### Strict Name Pattern Enforcement

If the provider behaviour of mangling with user's pattern is undesirable, it can be disabled. The `enforce` option allows you to strictly control whether providers can modify your specified naming patterns. When set to `true`, it prevents providers from applying their resource-specific naming rules or transformations.

The `enforce` option can be set at any level where patterns are specified:
```yaml
config:
  pulumi:autonaming:
    pattern: ${name}-${hex(8)}
    enforce: true    # Enforce this pattern globally
    providers:
      aws:
        pattern: ${name}_${hex(4)}
        enforce: true    # Enforce AWS-specific pattern
        resources:
          "aws:s3/bucket:Bucket": 
            pattern: ${name}${alphanum(6)}
            enforce: true    # Enforce bucket-specific pattern
```

When `enforce: true` is set, the provider will assign a generated name as-is and will attempt resource creation, even if the default autonaming rules don't allow it.

### Migration

Changing the autonaming setting on an existing stack doesn’t cause any immediate changes. It will only affect any newly created resources on this stack, including resources being replaced for unrelated reasons. To re-create resources with new names, e.g. on a dev stack, the user would need to destroy the stack and update it again.

## Implementation Proposal

The key task to solve is the negotiation of the resource name between the user's preferences set in the config and the provider’s knowledge of its cloud domain. There are two possibilities of which components makes the final decision:

1. **Provider**. User preferences are passed over to the provider and it builds the name based on the complete information.  
2. **Engine**. This involves schematizing of auto-naming rules by all providers, a (new) mechanism to load resource schemas in the engine, and then finally building the name based on all the information inside the engine and passing it over to the provider.

The proposal is to go ahead with option (1), for a few reasons:

1. Backwards compatibility. As mentioned above, most users likely will keep relying on the default behavior, so it’s important to keep it exactly the same. In this scenario, the engine won’t pass any new information to the provider and the provider will continue to do the same calculation it does today.  
2. Gradual upgrade. This approach doesn’t strictly require switching all providers to the new mechanism at once. It’s unlikely we can switch all the providers to the schema-based generation, so we’d end up in a mixed situation, which is harder to reason about. With provider-side generation, if a legacy provider does not respect the user's format, it’s probably not the end of the world. Conversely, provider-side name generation will need to stay operational indefinitely to support older CLI versions.  
3. Full flexibility. Providers still have full control over naming, so they can implement arbitrary cloud-specific rules when needed, including based on other resource inputs.  
4. Implementation simplicity. Fewer changes are needed across the stack, as described below.

### Include naming metadata in Check gRPC request

Extend the arguments to Check with optional autonaming metadata. The metadata are pre-calculated for the current provider and resource type.

### Calculate naming metadata in the engine

The CLI would read stack configuration and pass it down to the step generator. The step generator can calculate the exact naming template based on the resource’s URN.

It will use the seed value to calculate a “name candidate” that satisfies the user’s configuration and pass it over to the Check call. Centralizing name generation in the engine means we can evolve it more easily later on (e.g. add new template expressions).

### Respect naming metadata in each provider

We’ll then have to update each provider to respect the new parameter. Roughly, the logic should go:

1. If no parameter is supplied, do the same old calculation.  
2. If autonaming is disabled, do not generate any name automatically. Otherwise:
3. If there is a parameter supplied, use the “name candidate” as the name. 
4. If the `enforce` option is not enabled, for some exceptional resources like UUID-based ones, apply special rules to disambiguate.
5. If the `enforce` option is enabled, skip any special rules or validations.

In practice, this means:

* Change every native provider to respect the “name candidate” option, and apply item (4) where needed.  
* Change TF Bridge’s ComputeAutoNameDefault to respect the “name candidate” option. Also, implement item (4) above by applying Transform, MaxLength, Charset, Separator options to the “name candidate” to produce the final name or error (see [User preference vs. provider configuration](#user-preference-vs.-provider-configuration)).  
* Roll out the new bridge to all bridged providers.

### Version compatibility

If an older CLI invokes a provider with the new autonaming implementation, it won’t supply the name metadata and autonaming will work the default way.

If a new CLI invokes an older provider unaware of autonaming customization, the user’s preference will get ignored. The proposal is to treat it differently based on user’s configuration:

1. If no configuration is specified, all is good.  
2. If the user specified an explicit configuration for this provider (e.g. aws: ${name}-${hex(4)}), we should log a warning message and suggest upgrading to a newer provider version.  
3. If the user specified a global configuration (e.g. pulumi:autonaming: verbatim), we can silently ignore the mismatch and continue with legacy naming. The motivation for that is the long tail of providers that may not be upgradable or may not care about autonaming at all (e.g. random). We don’t want to flood the user with unactionable warnings.
4. If `enforce: true` is specified at any level applicable to the current resource, log a warning similar to (2).

For (2) and (4) above, we need to add a new feature support that the provider will report to the engine via a  SupportsFeature call.

### Create before replace

When a replacement is reported from Diff, the provider also reports the DeleteBeforeReplace flag value. Create-before-delete is only applicable if the resource names are generated with a random portion in its name. The exact generation mechanism of the replacement resource’s name isn’t important as long as it doesn’t produce the exact same name, which should never happen if the names are sufficiently random.

The engine should have a new condition while calculating DeleteBeforeReplace:

* If the user configured autonaming with no randomization (e.g. verbatim, or formatted with project/stack name and no random expressions), set DeleteBeforeReplace to true regardless of the value returned by the provider in Diff response.  
* Note that this will cause DeleteBeforeReplace in case autonaming configuration changes between the Create and Diff operations and the new configuration is not random, even though potentially the names could be different. This seems to be acceptable.  
* No changes to providers are required.

This seems to be the biggest danger for the users, and also the one they may not realize before they face a delete-before-create replacement. We should try to over-communicate this impact, e.g.

* Mention this impact in the docs for autonaming  
* Consider a one-time CLI warning on the first run after the setting is changed  
* Consider a CLI enhancement to more clearly indicate delete-before-create

#### Oct 23, 2024

- Added `disabled` mode
- Added `config.key` template
- Added JSON schema for the config shape definition
- Refactored the shape of autonaming configuration to be more explicit

#### Oct 31, 2024

- Added `enforce` option
- Refactored JSON schema to reference a reusable type

#### Dec 5, 2024

- Changed `${org}` expression to `${organization}` in line with current YAML language syntax

**Comment by lukehoban** [comment]
> `pulumi:autonaming`

The proposal uses one top level config key with three different interpretations:
1. A known magic value (`verbatim`, `default`)
2. An interpolated naming pattern (`${name}_${hex(4)}`)
3. A map from providers to (1) or (2) above.

I wonder if it might be better to separate these into different keys, or into a consistently typed schema under a single key?  For example, have `mode`, `pattern` and `providers` top level keys under `pulumi:autonaming`?


> ```
> pulumi:autonaming:
>   aws: ${name}_${hex(4)}   # AWS - use underscore and shorter suffix
>   azure-native:
>     default: verbatim      # disable name mangling for Azure
>     storage:Account: ${name}${string(6)}  # storage accounts are globally
>                                           # unique, so add a suffix
> ```


Similarly, I wonder if we want to be more consistent between the top level and nested schemas here?  For example, if we want a `default` key and an arbitrary set of type names, should they:
1. Use full type tokens consistently even in this nested location
2. Allow the same per-type-token approach at top level?
3. Seperate the magic `default` value into a different layer of the object.

Maybe more broadly, would be useful to see the whole schema for this configuration object, and ensure it can be "simple" while still accomplishing all the configurability desired here.

> When users provide a custom autonaming template, we want to keep that extra provider knowledge and apply it intelligently. Otherwise, users would have to encode all the same rules in their configurations.

Are we sure we want this?  What's the most compelling place we feel we need it?  I feel like its actually a good thing to let users have final control over overriding what the provider "thinks" is right, just in case there is some bug/limitation in the provider they need to workaround.  It's not clear that we shouldn't just expect users who want to override the behaviour to be explicit about the pattern they want?  

> All the ${exp} expressions are evaluated at the time of name generation, with the following expressions supported:

Curious - does this list cover all of the ways we currently customize autonaming in providers today?  Are there any esoteric autonaming overrides that are missing?

Related to this - how are name length restrictions handled?  I believe today autonaming has the ability to require the name to be no longer than N characters, and will take a prefix of the seed name before adding randomness to ensure this.  How do we see this working along with the patterns?  It seems they may frequently result in names that can be too long unless we allow users to force length restrictions?  Perhaps this is the motivation for the point above?  (Though I'm not sure by itself that would address this?)

**Reply by mikhailshilkov** [reply]
> would be useful to see the whole schema for this configuration object

Yes, makes sense, I'll work on adding a full schema.

> Are we sure we want this? What's the most compelling place we feel we need it?

The most compelling place is probably resources with a fixed name format - the last three bullet points of "known deviations": UUIDs, required fixed suffixes/prefixes, or path names. E.g. users with a broad Azure Native naming template won't be able to create policy assignments unless they add an explicit rule for that resource that maps to `${UUID}` (with no other valid option).

>  in case there is some bug/limitation in the provider they need to workaround

Do we know of existing cases like that?

> Curious - does this list cover all of the ways we currently customize autonaming in providers today? Are there any esoteric autonaming overrides that are missing?

I think it does - maybe except transformations (mainly, `toLower`). Providers could continue to apply transformations, as described in the proposal. Alternatively, we either need to extend the expressions with transformation "functions" or allow configuring naming in code.

> Related to this - how are name length restrictions handled?

This is described in the doc:

> The provider should error for cases when a name transformation can dangerously reduce randomness:
> - Where the generated name is longer than maximum length allowed (same as today)

I'm not sure how this changes between now and user configuration. Currently, we fail if the length is exceeded:
https://github.com/pulumi/pulumi/blob/fd07f42c6116ea9685b24ca35ba5832994196fc5/sdk/go/common/resource/resource_id.go#L102-L105

The same will happen for a user's template: provider will error complaining about the name length. Users will have to define a shorter logical name, or a different template for that particular resource type.

**Reply by mikhailshilkov** [reply]
> I wonder if we want to be more consistent between the top level and nested schemas here?

Adjusted the proposal in this direction.

> would be useful to see the whole schema for this configuration object, and ensure it can be "simple" while still accomplishing all the configurability desired here.

Added the JSON schema.

**Comment by DominicRoyStang** [comment]
Would this proposal allow for programmatically defining the naming behavior when instantiating a custom provider?

eg. something like this

```typescript
function customAutonaming(name: string) {
    return `${process.env.RESOURCE_PREFIX}-${name.toLowerCase()}-customSuffix`
}

const awsProvider = new aws.Provider('customProvider', { autonaming: customAutonaming }) 
```

**Reply by gunzy83** [reply]
> Would this proposal allow for programmatically defining the naming behavior when instantiating a custom provider?
> 
> eg. something like this
> 
> ```ts
> function customAutonaming(name: string) {
>     return `${process.env.RESOURCE_PREFIX}-${name.toLowerCase()}-customSuffix`
> }
> 
> const awsProvider = new aws.Provider('customProvider', { autonaming: customAutonaming }) 
> ```

We have already developed a set of naming helper functions as a part of a private package within our org, so something like this would work really well to enforce a naming standard.

**Reply by mikhailshilkov** [reply]
@DominicRoyStang @gunzy83 The current version of the proposal does not accommodate for programmatic naming (see "Non-Goals"). It's meaningfully harder to implement, so we'd want to ensure there is enough value added.

I'd be eager to learn more about the use cases why you want to do that programmatically:

1. Why do you prefer a function over a config value? Could both prefix and suffix be set via config values, or is there extra flexibility that I am not seeing?
2. Why do you want to set it per explicit provider? Will you have different naming conventions for different instances of the AWS provider, in your example?

> a set of naming helper functions

Could you please share more details about their variability? How do you apply functions from that set today, and how would you want to apply them?

**Reply by gunzy83** [reply]
> Why do you prefer a function over a config value?

The main reason is packaging and reuse. We have internal packages that do a heap of this work for you so you can import then call the relevant name templating function and we then don't need to check that in PRs, we can focus on the logic and architecture. Implementing this in config would work but the correct version of it would need to be copied into each project's `Pulumi.yaml`. Not a huge issue, just another thing to remember on new project creation rather than installing our internal packages (which automatically wire up many things so you want to install them).

> Could both prefix and suffix be set via config values, or is there extra flexibility that I am not seeing?
> Could you please share more details about their variability? How do you apply functions from that set today, and how would you want to apply them?

I thought I would answer these together as they are related. In our case this would be dynamic depending on where in our architecture the resource lives. What I mean by that is each stack is defined to "live" in a layer of the architecture which allows us to discover parent/ancestor and sibling stacks easily based on a common set of variables that form stack names (used in automagic `StackReference` construction, enforced when running the stack with our internal package), physical resource names, tags, K8s labels and Pagerduty services etc. The layers are:

* `global` - one stack across the org
* `account` - one stack per AWS account, the project name allows discovery of global stacks for references
* `regional` - one stack per region per AWS account, can discover account and global stacks
* `group` - infra group (eg VPC, cluster), many per region
* `environment` - many per infra group, typically application stacks running on the same cluster in different namespaces

We have a `prefix` variable that we use to allow us to break out additional stacks, so far mainly in the group level for multiple database instances in the same group but not specific to an environment.

The naming differs depending on the resource (mainly to aid in discovery in the Console/UI), for example:

IAM Roles:
* account level: `${role-purpose}`
* regional level: `${region}-${role-purpose}`
* group level, eg eks node role: `${region}-${group}-${role-purpose}`
* environment level, eg IRSA: `${region}-${group}-${environment}-${app-name}`

SQS/SNS:
* group level: `${group}-${purpose|application}`
* environment level: `${group}-${environment}-${purpose|application}`

Each implementation weighs the tradeoffs of using `namePrefix` or `name` (create before replace vs easier use in IAM and resource policies).

> The current version of the proposal does not accommodate for programmatic naming (see "Non-Goals"). It's meaningfully harder to implement, so we'd want to ensure there is enough value added.

After typing all of that, I agree it is meaningfully harder, and I would focus on the use cases that please the most of the users. Hopefully my extra info is useful.

**Comment by gunzy83** [comment]
> We have a design proposal for a lever to configure autonaming, or disable it entirely.

I can't see where in this proposal we can completely disable autonaming. I see `Disable name mangling` which will pass the logical name directly to the physical name, that is not disabling it in the sense of forcing the user to actually set the physical name (or name prefix argument) to meet organisational naming standards. 

At risk of sounding like a broken record, my thoughts on logical vs physical naming are here in the original issue: https://github.com/pulumi/pulumi/issues/1518#issuecomment-1696607694.

> ### Custom autonaming template
> ### Configure per provider or per resource type

Now these two are interesting. I would obviously not be in favour of using `${name}` within the template. Perhaps custom Pulumi config keys could be accessed also? This would make enforcing an organisational standard based on common stack variables we require in all of our stacks (which flow into tags and other fields) much easier.

**Reply by mikhailshilkov** [reply]
Yes, I can see us enabling both:
1. Add a `disable` option to auto-naming.
2. Add a `${config.foo}` expression to name templates.

Thank you for your suggestions!

**Reply by gunzy83** [reply]
Awesome. I can see myself using both.

**Comment by alexandre-ava** [comment]
Great work, that is a very interesting RFC that seem to solve the issues I currently have. Being able to configure autonaming by resource type is great. However, I am not sure I understood everything correctly concerning the "User preference vs. provider configuration" section.

If "Autonaming configured in explicit providers" is a non-goal, a provider like Azure-Native won't have any default logic or configuration to ensure the naming rules (listed [in the Azure documentation](https://learn.microsoft.com/en-us/azure/azure-resource-manager/management/resource-name-rules)) are respected and correctly applied. So I will probably have to set a big autonaming configuration by resource type to ensure the naming rules are applied, and I will have to do that for all my stacks.

Or maybe I did not understand correctly and the transformations rules described in "user preference vs. provider configuration" will be implemented in the providers, like the Azure-Native provider in my example.

**Reply by mikhailshilkov** [reply]
> the transformations rules described in "user preference vs. provider configuration" will be implemented in the providers, like the Azure-Native provider in my example.

The proposal does include this. The current idea is that the provider has the final decision and may decide to override the configuration provided by the user. Although, some folks are advocating against this approach and always to prefer the setting that users provided.

Let me ask your opinion with a concrete example. Let's say you configured some custom template:

```yaml
pulumi:autonaming:
  pattern: ${name}_${hex(6)}
```

but haven't customized anything per-provider or per-resource, to keep it simple. You use Azure Native.

0. When you create a Resource Group:

```ts
const rg = new azurenative.resources.ResourceGroup("myrg");
```

you get a name like `myrg_abcd12`. I think this is non-controversial.

1. Now, you want a Storage Account:

```ts
const sa = new azurenative.storage.StorageAccount("sa", { ... });
```

Here, the automatic name would be like `sa_abcd12`, which is not a valid storage account name. Would you want the provider to:
a. Remove the underscore for you and make it `saabcd12`?
b. Throw an error, making you define a separate setting for that resource type (or an explicit fixed name)?
c. Something else.

2. Next, you want a Blob in that account:

```ts
const blob = new azurenative.storage.Blob("my.zip", { ... });
```

The default way Azure Native handles it is to keep the "verbatim" name `my.zip` for the blob. But the user asked for a random suffix, broadly, so should the provider
a. Ignore the generated autoname and keep using `my.zip`?
b. Respect the generated autoname and create a blog `my.zip_abcd12`?
c. Something else?

3. Finally, you want to create a Policy Assignment:

```ts
const pa = new azurenative.authorization.PolicyAssignment("pa", { ... });
```

Assignment IDs are UUIDs. Do you want the provider to
a. Ignore the generated name and generate a UUID?
b. Error because the autonaming template is asking for a non-UUID?
c. Something else?

How do you answer (1), (2), and (3)? FWIW the proposal is 1a-2a-3a, but the downside is we end up not respecting user's preference.

Do your answers change if you configure a separate rule specific to one of resource types?

**Reply by TechWatching** [reply]
Thanks for the answer. With this example it's perfectly clear.

My answer would be similar to the proposal so :  1a (so `saabcd12`) - 2a - 3a.
Because:
- it's more convenient that the provider do the transformations than me having to specify all the rules for each resource type
- the provider knows better than me the naming rules to apply and I'm fine with that

If it's possible I would probably add warning or information logs in the stack update logs to explain why one of preference was not respected for some name.

Tough question, concerning the case where I have configured a specific rule on a resource type. I would be tempted to say 1b - 2b -3b to respect user preference (when a user set this kind of configuration, he is supposed to know what he is doing). But I could live with 1a -2a - 3a for that also because it might be more consistent to keep the same reasoning everywhere and not change depending on the configuration location.

I can understand why some folds would prefer their choice to be respected and not overidden by the provider. Maybe a 4th mode is needed ("preserveUser" would be a very bad name but you get the idea) to keep autonaming enabled but prevent the provider to override the user preference. With this mode, the answer to your example would be 1a - 2a - 3a when the mode is set to this 4th option.

**Reply by mikhailshilkov** [reply]
Great idea about `preserveUser `. I added an `enforce` option to the proposal, so that `(a)` is the default behaviour but users can opt into `(b)` with `enforce: true` on any level.

**Comment by olafurnielsen** [comment]
Really like this proposal. Pitching in our current implementation for reference.

In our (Azure focused) IaC library used across multiple customers we implemented autonaming based on naming recommendations from the Azure Cloud Adoption Framework:

- [Define your naming convention](https://learn.microsoft.com/en-us/azure/cloud-adoption-framework/ready/azure-best-practices/resource-naming)
- [Abbreviation recommendations for Azure resources](https://learn.microsoft.com/en-us/azure/cloud-adoption-framework/ready/azure-best-practices/resource-abbreviations)

In essence, the recommended naming format is:
`{resource abbreviation}-{workload name}-{environment}-{region}-{instance suffix}`

In praxis, we've implemented some custom logic to take care of this:
```python
resources.ResourceGroup(
  resource_name=stack.name_for(resources.ResourceGroup)
  ...
)
```

Which will generate a name by using:
- Resource abbreviation fetched from a local dictionary map. The list from Azure docs is not complete so we add prefixes for those missing.
- `workload_name` from stack config or Pulumi project name
- `env` from stack config or Pulumi stack name
- Abbreviation of the region configured (azure-native:location), i.e. northeurope -> neu
- Instance suffix: pulumi autosuffix hex for most resources but explicit for resources which have short naming restrictions

Resulting in the logical name:
`rg-projectname-prod-neu` with Pulumi random suffix added in the physical name.

**Is my understanding correct that the proposed implementation would**
- Allow prefixes based on resource type to be defined: 
  a. on the provider level (i.e. `azure-native`)
  b. via the explicit "per resource type" method
- Support OR operators, in our case: `${config.workload_name | project}` and `${config.env | stack}`?
- Not support custom dynamic logic, f.x. for abbreviations `${location_abbr(config.azure-native.location)}`? 
- Seems like most random suffix scenarios are supported. Except for autoincreasing instance counters (01, 02 ...) which I think is fine.

Implementing this was a major time saver for us so having this functionality as part of Pulumi would be quite awesome 🙌

**Reply by mikhailshilkov** [reply]
Thank you for describing your approach, and for your questions!

> Allow prefixes based on resource type to be defined:
> a. on the provider level (i.e. azure-native)

Yes, you'll be able to set a default prefix for all resources of that provider.

> b. via the explicit "per resource type" method

Yes, those would override the default provider-level prefix above.

> Support OR operators, in our case: ${config.workload_name | project} and ${config.env | stack}?

No, logic operators are outside of the current scope.

> Not support custom dynamic logic, f.x. for abbreviations ${location_abbr(config.azure-native.location)}?

Correct, custom functions are outside the scope. You'd have to pre-calculate the value and set it as a config or an environment variable.

> Seems like most random suffix scenarios are supported. Except for autoincreasing instance counters (01, 02 ...) which I think is fine.

Correct.

**Reply by olafurnielsen** [reply]
Despite these limitations in the initial scope, this would be a major quality of life improvement. Thanks for all the good work 🙏🏻

**Comment by mikhailshilkov** [comment]
This feature has now officially shipped! You can learn more about it in my blog post: [Introducing Customizable Resource Auto-naming in Pulumi](https://www.pulumi.com/blog/autonaming-configuration/)

Thank you to everyone who provided feedback and helped shape the proposal!

**Reply by stfnzl** [reply]
This is gr8! Thank for the effort!
